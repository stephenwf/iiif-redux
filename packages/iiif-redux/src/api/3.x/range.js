import memoize from 'lodash.memoize';
import { createSelector } from 'reselect';
import mapById from '../../utility/mapById';
import * as technical from './iiif/technical';
import * as descriptive from './iiif/descriptive';
import * as linking from './iiif/linking';
import * as structural from './iiif/structural';
import {
  getAllAnnotationCollections,
  getAllAnnotationPages,
  getAllAnnotations,
  getAllCanvases,
  getAllCanvasReferences,
  getAllContentResources,
  getAllRanges,
  getAllResources,
  getAllSelectors,
  getAllServices,
} from '../all';
import mapByIdOrId from '../../utility/mapByIdOrId';
import mapAllResources from '../../utility/mapAllResources';
import mapAllById from '../../utility/mapAllById';
import { standardFieldMappingFactory } from '../../../es/utility/new/standardFieldMappingFactory';

const range = memoize(selector => {
  /**
   * Technical properties
   *
   * - getId
   * - getType
   * - getBehavior
   * - getViewingDirection
   */
  const getId = createSelector(selector, technical.getId);

  const getType = createSelector(selector, technical.getType);

  const getViewingDirection = createSelector(
    selector,
    technical.getViewingDirection
  );

  const getBehavior = createSelector(selector, technical.getBehavior);

  /**
   * Descriptive Properties
   *
   * - getLabel
   * - getMetadata
   * - getSummary
   * - getThumbnail
   * - getPosterCanvas
   * - getRequiredStatement
   * - getRights
   * - getNavDate
   */
  const getLabel = createSelector(selector, descriptive.getLabel);

  const getMetadata = createSelector(selector, descriptive.getMetadata);

  const getSummary = createSelector(selector, descriptive.getSummary);

  const getThumbnailId = createSelector(selector, descriptive.getThumbnailId);

  const getThumbnail = createSelector(
    getThumbnailId,
    getAllContentResources,
    mapByIdOrId
  );

  const getPosterCanvasId = createSelector(
    selector,
    descriptive.getPosterCanvas
  );
  const getPosterCanvas = createSelector(
    getPosterCanvasId,
    getAllCanvases,
    mapById
  );

  const getRequiredStatement = createSelector(
    selector,
    descriptive.getRequiredStatement
  );

  const getRights = createSelector(selector, descriptive.getRights);

  const getNavDate = createSelector(selector, descriptive.getNavDate);

  /**
   * Linking properties
   *
   * - getSeeAlso
   * - getService
   * - getLogo
   * - getHomepage
   * - getRendering
   * - getPartOf
   */
  const getSeeAlsoIds = createSelector(selector, linking.getSeeAlso);
  const getSeeAlso = createSelector(
    getSeeAlsoIds,
    getAllContentResources,
    mapAllById
  );

  const getServiceIds = createSelector(selector, linking.getService);
  const getService = createSelector(getServiceIds, getAllServices, mapAllById);

  const getLogoIds = createSelector(selector, linking.getLogo);
  const getLogo = createSelector(
    getLogoIds,
    getAllContentResources,
    mapAllById
  );

  const getHomepageId = createSelector(selector, linking.getHomepage);
  const getHomepage = createSelector(
    getHomepageId,
    getAllContentResources,
    mapById
  );

  const getRenderingIds = createSelector(selector, linking.getRendering);
  const getRendering = createSelector(
    getRenderingIds,
    getAllContentResources,
    mapAllById
  );

  const getPartOfId = createSelector(selector, linking.getPartOf);
  const getPartOf = createSelector(
    getPartOfId,
    getAllResources,
    mapAllResources
  );

  const getStartId = createSelector(selector, linking.getStart);
  const getStart = createSelector(
    getStartId,
    getAllCanvases,
    getAllCanvasReferences,
    (start, allCanvases, allCanvasReferences) => {
      if (!start) {
        return null;
      }
      if (start.schema === 'canvas') {
        return allCanvases[start.id];
      }
      return allCanvasReferences[start.id] || null;
    }
  );

  const getSupplementaryId = createSelector(selector, linking.getSupplementary);
  const getSupplementary = createSelector(
    getSupplementaryId,
    getAllAnnotationCollections,
    mapById
  );

  /**
   * Structural Properties
   *
   * - getItems
   * - getRanges
   * - getAnnotations
   */
  const getItemIds = createSelector(selector, structural.getItems);
  const getItems = createSelector(
    getItemIds,
    getAllAnnotations,
    getAllCanvases,
    getAllRanges,
    getAllResources,
    getAllCanvasReferences,
    getAllSelectors,
    (
      itemIds,
      allAnnotations,
      allCanvases,
      allRanges,
      allResources,
      allCanvasReferences,
      allSelectors
    ) => {
      const map = {
        annotation: allAnnotations,
        canvas: allCanvases,
        range: allRanges,
        resource: allResources,
        canvasReference: allCanvasReferences,
        selector: allSelectors,
      };
      return itemIds.map(({ schema, id }) => map[schema][id]);
    }
  );

  const getRangeIds = createSelector(getItemIds, rangeIds =>
    rangeIds
      .filter(({ schema }) => schema === 'range')
      .map(singleRange => singleRange.id)
  );
  const getRanges = createSelector(getRangeIds, getAllRanges, mapAllById);

  const getAnnotationIds = createSelector(selector, structural.getAnnotations);
  const getAnnotations = createSelector(
    getAnnotationIds,
    getAllAnnotationPages,
    mapAllById
  );

  return {
    // Technical.
    getId,
    getType,
    getBehavior,
    getViewingDirection,

    // Descriptive.
    getLabel,
    getMetadata,
    getSummary,
    getThumbnailId,
    getThumbnail,
    getPosterCanvasId,
    getPosterCanvas,
    getRequiredStatement,
    getRights,
    getNavDate,

    // Linking
    getSeeAlsoIds,
    getSeeAlso,
    getServiceIds,
    getService,
    getLogoIds,
    getLogo,
    getHomepageId,
    getHomepage,
    getRenderingIds,
    getRendering,
    getPartOfId,
    getPartOf,
    getStartId,
    getStart,
    getSupplementaryId,
    getSupplementary,

    // Structural.
    getItemIds,
    getItems,
    getRangeIds,
    getRanges,
    getAnnotationIds,
    getAnnotations,
  };
});

export const mappings = standardFieldMappingFactory(range);

export default range;
