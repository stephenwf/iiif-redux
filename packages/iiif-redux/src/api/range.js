import resourceListSelectorFactory from '../utility/resourceListSelectorFactory';
import byIdSelectorFactory from '../utility/byIdSelectorFactory';
import { getAllRanges } from './all';
import presentation2Range from './2.x/range';

const range = presentation2Range;

export default range;

export const ranges = resourceListSelectorFactory(getAllRanges, range);

export const rangeByIdSelector = byIdSelectorFactory(range, 'ranges');
