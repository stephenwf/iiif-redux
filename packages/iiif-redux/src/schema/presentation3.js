import { schema, normalize } from 'normalizr';
import { compose } from 'redux';
import addMissingIds from '../compat/addMissingIds';

function createEntity(name) {
  const options = {
    idAttribute: 'id',
  };

  return new schema.Entity(name, {}, options);
}

const collection = createEntity('collections'); // 1
const manifest = createEntity('manifests'); // 2
const canvas = createEntity('canvases'); // 3
const annotation = createEntity('annotations'); // 4
const annotationPage = createEntity('annotationPages'); // 5
const annotationCollection = createEntity('annotationCollection'); // 6
const range = createEntity('ranges'); // 7
const contentResource = createEntity('contentResources'); // 8
const choice = createEntity('choices'); // 9
const canvasReference = createEntity('canvasReference'); // 10

// Unofficial types.
const externalResource = createEntity('externalResources');
const service = createEntity('services');

// Union types
const partOf = new schema.Array(
  {
    collection,
    manifest,
    externalResource,
    annotationCollection,
  },
  entity => {
    switch (entity.type) {
      case 'Collection':
        return 'collection';
      case 'Manifest':
        return 'manifest';
      case 'AnnotationCollection':
        return 'annotationCollection';
      default:
        return 'externalResource';
    }
  }
);

const canvasOrReference = new schema.Union(
  {
    canvasReference,
    canvas,
  },
  entity => {
    if (entity.type === 'SpecificResource') {
      return 'canvasReference';
    }
    return 'canvas';
  }
);

const annotationBody = new schema.Union(
  {
    contentResource,
    externalResource,
    choice,
  },
  input => {
    switch (input.type) {
      case 'Application':
      case 'Dataset':
      case 'Image':
      case 'Sound':
      case 'Text':
      case 'Video':
      case 'TextualBody':
        return 'contentResource';
      case 'Choice':
        return 'choice';

      // @todo more content to add.
      default:
        return 'externalResource';
    }
  }
);

const rangeItem = new schema.Union(
  {
    canvas,
    range,
    canvasReference,
  },
  input => {
    switch (input.type) {
      case 'Canvas':
        return 'canvas';
      case 'Range':
        return 'range';
      case 'SpecificResource':
      default:
        // @todo default?
        return 'canvasReference';
    }
  }
);

// Root resource
/**
 * Dereferencable resources:
 *
 * - Collection (Required)
 * - Manifest (Required)
 * - Canvas (Recommended)
 * - Annotation (Recommended)
 * - AnnotationPage (Required)
 * - AnnotationCollection (Optional)
 * - Content Resource (Required)
 * - Range (Optional)
 */
const RESOURCE_TYPE_MAP = {
  Collection: 'collection',
  Sequence: 'sequence',
  Manifest: 'manifest',
  Canvas: 'canvas',
  AnnotationCollection: 'annotationCollection',
  AnnotationPage: 'annotationPage',
  Annotation: 'annotation',
  Range: 'range',
  // Content resources.
  Application: 'contentResource',
  Dataset: 'contentResource',
  Image: 'contentResource',
  Sound: 'contentResource',
  Text: 'contentResource',
  Video: 'contentResource',
  TextualBody: 'contentResource',
};
const resource = new schema.Union(
  {
    collection,
    manifest,
    canvas,
    annotationCollection,
    annotationPage,
    annotation,
    range,
    contentResource,
    service,
  },
  entity => {
    if (RESOURCE_TYPE_MAP[entity.type]) {
      return RESOURCE_TYPE_MAP[entity.type];
    }
    if (entity.profile) {
      return 'service';
    }
    throw Error('Unknown entity type');
  }
);

// ===========================================================================
// 1) Collections
// ===========================================================================
collection.define({
  // Structural
  items: new schema.Array(
    {
      collection,
      manifest,
    },
    entity => (entity.type === 'Manifest' ? 'manifest' : 'collection')
  ),
  annotations: [annotationPage],

  // Linking.
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // Extra.
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 2) Manifests
// ===========================================================================
manifest.define({
  items: [canvas],
  structures: [range],
  annotations: [annotationPage],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,
  start: canvasOrReference,

  // Extra
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 3) Canvases
// ===========================================================================
canvas.define({
  // Structural.
  items: [annotationPage],
  annotations: [annotationPage],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],
  posterCanvas: canvas,
});

// ===========================================================================
// 4) Annotation
// ===========================================================================
annotation.define({
  // Structural.
  body: annotationBody,
  target: canvas,

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],
});

// ===========================================================================
// 5) Annotation Collection
// ===========================================================================
annotationCollection.define({
  // Structural.
  items: [annotationPage],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // extra
  thumbnail: [contentResource],

  // Paging.
  first: annotationPage,
  last: annotationPage,
});

// ===========================================================================
// 6) Annotation Page
// ===========================================================================
annotationPage.define({
  // Structural.
  items: [annotation],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,

  // Extra
  thumbnail: [contentResource],

  // Paging.
  next: annotationPage,
  previous: annotationPage,
});

// ===========================================================================
// 7) Range
// ===========================================================================
range.define({
  // Structural.
  items: [rangeItem],

  // Linking
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  homepage: externalResource,
  rendering: [externalResource],
  partOf: partOf,
  start: canvasOrReference,
  supplementary: annotationCollection,

  // Extra
  thumbnail: [contentResource],
});

// ===========================================================================
// 8) Content Resource
// ===========================================================================
contentResource.define({
  // Linking.
  seeAlso: [externalResource],
  service: [service],
  logo: [contentResource],
  rendering: [externalResource],
  partOf: partOf,
});

// ===========================================================================
// 9) Choice
// ===========================================================================
choice.define({
  default: annotation,
  item: [annotation],
});

// ===========================================================================
// 10) Canvas Reference
// ===========================================================================
canvasReference.define({
  source: canvas,
});

const preprocess = compose(addMissingIds(3.0));

const normalizeResource = (rawResource, customSchema = resource) =>
  normalize(preprocess(rawResource), customSchema);

export {
  collection,
  manifest,
  canvas,
  annotation,
  annotationPage,
  annotationCollection,
  range,
  contentResource,
  choice,
  canvasReference,
  externalResource,
  service,
  partOf,
  annotationBody,
  rangeItem,
  RESOURCE_TYPE_MAP,
  resource,
  preprocess,
  normalizeResource as normalize,
};
