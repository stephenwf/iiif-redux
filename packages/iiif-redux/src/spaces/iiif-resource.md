# IIIF Resources
This state is where all of our entities get imported into. The actions here are dynamic
and will call off to other spaces with follow up actions.

## State structure
```js
const state = {
  dereferenced: {
    'http://.../collection.json': { 
      loading: true,
      requested: '00-00-0000:00:00:00', 
      ttl: 600,
      error: null, // or { message: 'Failed to fetch', ... }
    },
  },
  collections: {
    'http://.../collection.json': {/* ... */},
  },
  manifests: {},
  sequences: {},
  canvass: {},
  annotations: {},
  annotationLists: {},
  ranges: {},
  layers: {},
  imageResources: {},
  externalResources: {},
  services: {},
}
```

## Reusable actions
Requesting an item that can definitely be referenced (i.e. manifest):
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    schema: schemas.manifest,
  }
});
```

Requesting an items fields, dereferencing if they don't exist.
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    dereferenceIfMissing: ['sequences'],
    schema: schemas.imageService,
  }
});
```

Requesting a resource, with a TTL so if its requested again a fresh copy may be requested.
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    ttl: 600,
    schema: schemas.manifest,
  }
});
```

Force-requesting an item, includes cache busting
```js
dispatch({
  type: 'REQUEST_IIIF_RESOURCE',
  payload: {
    resource: 'http://.../manifest.json',
    types: ['MANIFEST_REQUEST', 'MANIFEST_SUCCESS', 'MANIFEST_ERROR'],
    ttl: 600,
    forceFresh: true,
    schema: schemas.manifest,
  }
});
```

## Helper selectors
For use in containers for a more verbose way of showing resources current state while requesting.
```js
isResourceLoaded(id); // returns true/false, for loading screens.
didResourceError(id); // returns error object or null
```

## Potential sagas
In each of the spaces for the different resource types, we can have different reactions and
store different state. 

```js
function* manifestSaga() {
  yield takeEvery('MANIFEST_SUCCESS', handleSuccess);
} 

function* handleSuccess({ payload: { id } }) {
  // Set the current manifest to the one we just requested.
  yield put(setCurrentManifest(id));
  
  // Grab the state + start canvas
  const state = yield select();
  const startCanvas = getStartCanvas(state);

  // Set the canvas if its set.
  if (startCanvas) {
    yield put(setCurrentCanvasIndex(startCanvas)); // Another action creator.
  }
}
```
