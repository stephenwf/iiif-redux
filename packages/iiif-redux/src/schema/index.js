import * as presentation2 from './presentation2';
import * as presentation3 from './presentation3';

export { presentation2, presentation3 };

// Default is 2 until 3 is ratified?
export default presentation2;
