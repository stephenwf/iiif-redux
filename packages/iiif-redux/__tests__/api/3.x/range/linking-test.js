import range from '../../../../src/api/3.x/range';
import { normalize } from '../../../../src/schema/presentation3';
import fixtures from './fixtures';
import { DEFAULT_STATE } from '../../../../src/spaces/iiif-resource';

describe('api/3.x/range/linking', () => {
  describe.each(fixtures)('Fixture: %s', (name, fixture) => {
    const {
      result: { id },
      entities,
    } = normalize(fixture);
    const state = {
      resources: { ...DEFAULT_STATE.resources, ...entities },
    };
    const selectRange = s => s.resources.ranges[id];

    const {
      getSeeAlsoIds,
      getSeeAlso,
      getServiceIds,
      getService,
      getLogoIds,
      getLogo,
      getHomepageId,
      getHomepage,
      getRenderingIds,
      getRendering,
      getPartOfId,
      getPartOf,
      getStartId,
      getStart,
      getSupplementaryId,
      getSupplementary,
    } = range(selectRange);

    test('getSeeAlsoIds', () => {
      expect(getSeeAlsoIds(state)).toMatchSnapshot('getSeeAlsoIds');
    });

    test('getSeeAlso', () => {
      expect(getSeeAlso(state)).toMatchSnapshot('getSeeAlso');
    });

    test('getServiceIds', () => {
      expect(getServiceIds(state)).toMatchSnapshot('getServiceIds');
    });

    test('getService', () => {
      expect(getService(state)).toMatchSnapshot('getService');
    });

    test('getLogoIds', () => {
      expect(getLogoIds(state)).toMatchSnapshot('getLogoIds');
    });

    test('getLogo', () => {
      expect(getLogo(state)).toMatchSnapshot('getLogo');
    });

    test('getHomepageId', () => {
      expect(getHomepageId(state)).toMatchSnapshot('getHomepageId');
    });

    test('getHomepage', () => {
      expect(getHomepage(state)).toMatchSnapshot('getHomepage');
    });

    test('getRenderingIds', () => {
      expect(getRenderingIds(state)).toMatchSnapshot('getRenderingIds');
    });

    test('getRendering', () => {
      expect(getRendering(state)).toMatchSnapshot('getRendering');
    });

    test('getPartOfId', () => {
      expect(getPartOfId(state)).toMatchSnapshot('getPartOfId');
    });

    test('getPartOf', () => {
      expect(getPartOf(state)).toMatchSnapshot('getPartOf');
    });

    test('getStartId', () => {
      expect(getStartId(state)).toMatchSnapshot('getStartId');
    });

    test('getStart', () => {
      expect(getStart(state)).toMatchSnapshot('getStart');
    });

    test('getSupplementaryId', () => {
      expect(getSupplementaryId(state)).toMatchSnapshot('getSupplementaryId');
    });

    test('getSupplementary', () => {
      expect(getSupplementary(state)).toMatchSnapshot('getSupplementary');
    });
  });
});
