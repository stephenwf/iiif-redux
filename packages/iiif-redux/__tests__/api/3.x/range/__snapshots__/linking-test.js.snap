// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api/3.x/range/linking Fixture: Full range getHomepage: getHomepage 1`] = `
Object {
  "format": "text/html",
  "id": "https://example.org/info/book1/",
  "label": Object {
    "en": Array [
      "Home page for Book 1",
    ],
  },
  "type": "Text",
}
`;

exports[`api/3.x/range/linking Fixture: Full range getHomepageId: getHomepageId 1`] = `"https://example.org/info/book1/"`;

exports[`api/3.x/range/linking Fixture: Full range getLogo: getLogo 1`] = `
Array [
  Object {
    "id": "https://example.org/logos/institution1.jpg",
    "service": Array [
      "https://example.org/service/inst1",
    ],
    "type": "Image",
  },
]
`;

exports[`api/3.x/range/linking Fixture: Full range getLogoIds: getLogoIds 1`] = `
Array [
  "https://example.org/logos/institution1.jpg",
]
`;

exports[`api/3.x/range/linking Fixture: Full range getPartOf: getPartOf 1`] = `
Array [
  Object {
    "id": "https://example.org/collections/books/",
    "type": "Collection",
  },
]
`;

exports[`api/3.x/range/linking Fixture: Full range getPartOfId: getPartOfId 1`] = `
Array [
  Object {
    "id": "https://example.org/collections/books/",
    "schema": "collection",
  },
]
`;

exports[`api/3.x/range/linking Fixture: Full range getRendering: getRendering 1`] = `
Array [
  Object {
    "format": "application/pdf",
    "id": "https://example.org/iiif/book1.pdf",
    "label": Object {
      "en": Array [
        "Download as PDF",
      ],
    },
    "type": "Text",
  },
]
`;

exports[`api/3.x/range/linking Fixture: Full range getRenderingIds: getRenderingIds 1`] = `
Array [
  "https://example.org/iiif/book1.pdf",
]
`;

exports[`api/3.x/range/linking Fixture: Full range getSeeAlso: getSeeAlso 1`] = `
Array [
  Object {
    "format": "text/xml",
    "id": "https://example.org/library/catalog/book1.xml",
    "profile": "https://example.org/profiles/bibliographic",
    "type": "Dataset",
  },
]
`;

exports[`api/3.x/range/linking Fixture: Full range getSeeAlsoIds: getSeeAlsoIds 1`] = `
Array [
  "https://example.org/library/catalog/book1.xml",
]
`;

exports[`api/3.x/range/linking Fixture: Full range getService: getService 1`] = `
Array [
  Object {
    "id": "https://example.org/service/example",
    "profile": "https://example.org/docs/example-service.html",
    "type": "Service",
  },
]
`;

exports[`api/3.x/range/linking Fixture: Full range getServiceIds: getServiceIds 1`] = `
Array [
  "https://example.org/service/example",
]
`;

exports[`api/3.x/range/linking Fixture: Full range getStart: getStart 1`] = `
Object {
  "id": "https://example.org/iiif/book1/canvas/p2",
  "type": "Canvas",
}
`;

exports[`api/3.x/range/linking Fixture: Full range getStartId: getStartId 1`] = `
Object {
  "id": "https://example.org/iiif/book1/canvas/p2",
  "schema": "canvas",
}
`;

exports[`api/3.x/range/linking Fixture: Full range getSupplementary: getSupplementary 1`] = `
Object {
  "id": "https://example.org/iiif/1/annos/1",
  "type": "AnnotationCollection",
}
`;

exports[`api/3.x/range/linking Fixture: Full range getSupplementaryId: getSupplementaryId 1`] = `"https://example.org/iiif/1/annos/1"`;

exports[`api/3.x/range/linking Fixture: Range from IIIF spec getHomepage: getHomepage 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range from IIIF spec getHomepageId: getHomepageId 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range from IIIF spec getLogo: getLogo 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range from IIIF spec getLogoIds: getLogoIds 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range from IIIF spec getPartOf: getPartOf 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range from IIIF spec getPartOfId: getPartOfId 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range from IIIF spec getRendering: getRendering 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range from IIIF spec getRenderingIds: getRenderingIds 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range from IIIF spec getSeeAlso: getSeeAlso 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range from IIIF spec getSeeAlsoIds: getSeeAlsoIds 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range from IIIF spec getService: getService 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range from IIIF spec getServiceIds: getServiceIds 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range from IIIF spec getStart: getStart 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range from IIIF spec getStartId: getStartId 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range from IIIF spec getSupplementary: getSupplementary 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range from IIIF spec getSupplementaryId: getSupplementaryId 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range minimal getHomepage: getHomepage 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range minimal getHomepageId: getHomepageId 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range minimal getLogo: getLogo 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range minimal getLogoIds: getLogoIds 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range minimal getPartOf: getPartOf 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range minimal getPartOfId: getPartOfId 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range minimal getRendering: getRendering 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range minimal getRenderingIds: getRenderingIds 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range minimal getSeeAlso: getSeeAlso 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range minimal getSeeAlsoIds: getSeeAlsoIds 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range minimal getService: getService 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range minimal getServiceIds: getServiceIds 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range minimal getStart: getStart 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range minimal getStartId: getStartId 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range minimal getSupplementary: getSupplementary 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range minimal getSupplementaryId: getSupplementaryId 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range with start property getHomepage: getHomepage 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range with start property getHomepageId: getHomepageId 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range with start property getLogo: getLogo 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range with start property getLogoIds: getLogoIds 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range with start property getPartOf: getPartOf 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range with start property getPartOfId: getPartOfId 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range with start property getRendering: getRendering 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range with start property getRenderingIds: getRenderingIds 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range with start property getSeeAlso: getSeeAlso 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range with start property getSeeAlsoIds: getSeeAlsoIds 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range with start property getService: getService 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range with start property getServiceIds: getServiceIds 1`] = `Array []`;

exports[`api/3.x/range/linking Fixture: Range with start property getStart: getStart 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range with start property getStartId: getStartId 1`] = `
Object {
  "id": "https://example.org/iiif/book1/canvas/p1#xywh=10,20,30,40",
  "schema": "selector",
}
`;

exports[`api/3.x/range/linking Fixture: Range with start property getSupplementary: getSupplementary 1`] = `null`;

exports[`api/3.x/range/linking Fixture: Range with start property getSupplementaryId: getSupplementaryId 1`] = `Array []`;
